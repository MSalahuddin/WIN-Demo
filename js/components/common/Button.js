import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/native';
import LinearGradient from 'react-native-linear-gradient';
import { color, buttonColor } from '../../styles';
import { scaleHeight, scaleWidth } from '../../platformUtils';
import { SOUNDS } from '../../soundUtils';
import { BackgroundMusicContext } from '../../context/BackgroundMusic.context';

export const BUTTON_COLOR_SCHEME = {
  GREEN: 'GREEN',
  GREY: 'GREY',
  NORMAL_INVERTED: 'NORMAL_INVERTED',
  NORMAL: 'NORMAL',
  PURPLE: 'PURPLE',
  RED: 'RED',
  WHITE: 'WHITE',
  BLUE: 'BLUE',
};
const ButtonContainer = styled.View`

  justify-content: center;
  align-items: center;
  margin-horizontal: ${scaleHeight(5)};
  margin-vertical: ${scaleWidth(5)};
 `;

const ButtonOuterView = styled.TouchableOpacity`
  align-items: center;
  ${({ colorScheme }) => colorScheme && `background-color : ${colorScheme}`};
  border-radius: ${({ borderRadius }) => scaleHeight(borderRadius)};
  height: ${({ height }) => scaleHeight(height)};
  justify-content: center;
  width: ${({ width }) => scaleWidth(width)};
 `;

const ButtonHighlightView = styled.View`
  align-items: center;
  background-color: ${({ colorScheme }) => colorScheme};
  border-radius: ${({ borderRadius }) => scaleHeight(borderRadius)};
  height: ${({ height }) => scaleHeight(height - 5)};
  justify-content: center;
  width: ${({ width }) => scaleWidth(width - 3)};
`;
const ButtonHighlightGradient = styled(LinearGradient)`
align-items: center;
border-radius: ${({ borderRadius }) => scaleHeight(borderRadius)};
height: ${({ height }) => scaleHeight(height)};
justify-content: center;
width: ${({ width }) => scaleWidth(width)};
`;

const ButtonInnerView = styled.View`
  align-items: center;
  ${({ colorScheme }) => colorScheme && `background-color : ${colorScheme}`};
  border-radius: ${({ borderRadius }) => scaleHeight(borderRadius + 10)};
  height: ${({ height }) => scaleHeight(height - 7)};
  justify-content: center;
  margin-top: ${scaleHeight(4)};
  width: ${({ width }) => scaleWidth(width - 3)};
`;
const ButtonInnerGradientView = styled(LinearGradient)`
  align-items: center;
  ${({ colorScheme }) => colorScheme && `border-color : red`};
  border-radius: ${({ borderRadius }) => scaleHeight(borderRadius + 10)};
  height: ${({ height }) => scaleHeight(height - 5)};
  justify-content: center;
  border-width:1;
  margin-top: ${scaleHeight(3)};
  width: ${({ width }) => scaleWidth(width - 1)};
`;

const Button = ({
  children,
  onPress,
  theme,
  height,
  disabled,
  width,
  borderRadius,
  onLayout,
  gardient,
  gradientColor,
  gardientHighlight,
  ...rest
}) => {
  const { playSoundEffect } = useContext(BackgroundMusicContext);
  const COLOR_SCHEME = buttonColor[theme];

  const playButtonSound = () => {
    if (width > scaleWidth(160)) {
      if (theme === BUTTON_COLOR_SCHEME.GREEN || theme === BUTTON_COLOR_SCHEME.BLUE) {
        playSoundEffect(SOUNDS.GREEN_BUTTON);
      } else {
        playSoundEffect(SOUNDS.BLUE_BUTTON);
      }
    } else {
      playSoundEffect(SOUNDS.MINOR_BUTTON);
    }
  };

  const handleOnPress = () => {
    playButtonSound();
    onPress();
  };
  return (
    <ButtonContainer
      onLayout={onLayout}

    >
      <ButtonOuterView
        testID="button"
        activeOpacity={disabled ? 1 : 0.5}
        onPress={() => {
          if (!disabled) {
            handleOnPress();
          }
        }}
        {...rest} colorScheme={gardient ? null : COLOR_SCHEME.outer}

        borderRadius={borderRadius} height={height} width={width}
      >
        {gardient ? <ButtonHighlightGradient
          colors={gradientColor}
          start={{ x: 0, y: 0 }} end={{ x: 1, y: 0 }}
          borderRadius={borderRadius}
          height={height}
          width={width}
        >
          <ButtonInnerView borderRadius={borderRadius} height={height} width={width}>
            {children}
          </ButtonInnerView>
        </ButtonHighlightGradient> :
          <ButtonHighlightView
            colorScheme={COLOR_SCHEME.highlight}
            borderRadius={borderRadius}
            height={height}
            width={width}
            gardientHighlight
          >
            {gardientHighlight ?
              <ButtonInnerGradientView colorScheme={COLOR_SCHEME.highlight} colors={gradientColor} borderRadius={borderRadius} height={height} width={width}>
                {children}
              </ButtonInnerGradientView>
              : <ButtonInnerView colorScheme={COLOR_SCHEME.inner} borderRadius={borderRadius} height={height} width={width}>
                {children}
              </ButtonInnerView>}
          </ButtonHighlightView>
        }
      </ButtonOuterView>
    </ButtonContainer>
  );
};

Button.defaultProps = {
  disabled: false,
  height: 55,
  borderRadius: 40,
  onPress: () => { },
  theme: BUTTON_COLOR_SCHEME.NORMAL,
  width: 160,
  onLayout: () => { },
  gardient: false,
  gardientHighlight: false,
  gradientColor: [color.fuchsiaBlue, color.heliotrope]
};

Button.propTypes = {
  onLayout: PropTypes.func,
  disabled: PropTypes.bool,
  borderRadius: PropTypes.number,
  children: PropTypes.node.isRequired,
  height: PropTypes.number,
  onPress: PropTypes.func,
  theme: PropTypes.oneOf([
    BUTTON_COLOR_SCHEME.GREEN,
    BUTTON_COLOR_SCHEME.GREY,
    BUTTON_COLOR_SCHEME.NORMAL_INVERTED,
    BUTTON_COLOR_SCHEME.NORMAL,
    BUTTON_COLOR_SCHEME.PURPLE,
    BUTTON_COLOR_SCHEME.RED,
    BUTTON_COLOR_SCHEME.WHITE,
    BUTTON_COLOR_SCHEME.BLUE
  ]),
  width: PropTypes.number,
  gardient: PropTypes.bool,
  gardientHighlight: PropTypes.bool,
  gradientColor: PropTypes.node
};

export default Button;
