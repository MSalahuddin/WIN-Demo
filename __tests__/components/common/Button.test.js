import React from 'react';
import { render, fireEvent } from 'react-native-testing-library';

import Button from '../../../js/components/common/Button';
import MockProvider from '../../MockProvider';

describe('Component Button', () => {
  const onPress = jest.fn();
  const onPressFunction = jest.fn();

  test('renders regular button correctly', () => {
    const output = render(
      <MockProvider>
        <Button>{1}</Button>
      </MockProvider>
    );
    expect(output).toMatchSnapshot();
  });

  test('renders disabled button correctly', () => {
    const output = render(
      <MockProvider>
        <Button disabled>{1}</Button>
      </MockProvider>
    );
    expect(output).toMatchSnapshot();
  });

  test('calls onPress when button is pressed', () => {
    const { getByTestId } = render(
      <MockProvider>
        <Button onPress={onPress}>{1}</Button>
      </MockProvider>
    );
    fireEvent(getByTestId('button'), 'press');
    expect(onPress).toHaveBeenCalled();
  });

  test('not call onPress when button is disabled', () => {
    const { getByTestId } = render(
      <MockProvider>
        <Button disabled onPress={onPressFunction}>
          {1}
        </Button>
      </MockProvider>
    );
    fireEvent(getByTestId('button'), 'press');
    expect(onPressFunction).not.toHaveBeenCalled();
  });
});
