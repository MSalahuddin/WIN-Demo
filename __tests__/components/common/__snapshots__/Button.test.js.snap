// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Button renders disabled button correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginHorizontal": 10,
        "marginVertical": 10,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(116, 158, 253, 1)",
        "borderRadius": 80,
        "height": 110,
        "justifyContent": "center",
        "opacity": 1,
        "width": 320,
      }
    }
    testID="button"
  >
    <View
      borderRadius={40}
      colorScheme="rgba(165, 193, 255, 1)"
      height={55}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(165, 193, 255, 1)",
            "borderRadius": 80,
            "height": 100,
            "justifyContent": "center",
            "width": 314,
          },
        ]
      }
      width={160}
    >
      <View
        borderRadius={40}
        colorScheme="rgba(127, 166, 253, 1)"
        height={55}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(127, 166, 253, 1)",
              "borderRadius": 100,
              "height": 96,
              "justifyContent": "center",
              "marginTop": 8,
              "width": 314,
            },
          ]
        }
        width={160}
      >
        1
      </View>
    </View>
  </View>
</View>
`;

exports[`Component Button renders regular button correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginHorizontal": 10,
        "marginVertical": 10,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(116, 158, 253, 1)",
        "borderRadius": 80,
        "height": 110,
        "justifyContent": "center",
        "opacity": 1,
        "width": 320,
      }
    }
    testID="button"
  >
    <View
      borderRadius={40}
      colorScheme="rgba(165, 193, 255, 1)"
      height={55}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(165, 193, 255, 1)",
            "borderRadius": 80,
            "height": 100,
            "justifyContent": "center",
            "width": 314,
          },
        ]
      }
      width={160}
    >
      <View
        borderRadius={40}
        colorScheme="rgba(127, 166, 253, 1)"
        height={55}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(127, 166, 253, 1)",
              "borderRadius": 100,
              "height": 96,
              "justifyContent": "center",
              "marginTop": 8,
              "width": 314,
            },
          ]
        }
        width={160}
      >
        1
      </View>
    </View>
  </View>
</View>
`;
